File: Aula01-introducao-poo/ts/ParkingTicket.ts
export class ParkingTicket {
  private readonly entryTime: Date;
  private exitTime?: Date;
  public static readonly HOURLY_RATE = 6.0;

  constructor(entryTime: Date) {
    if (!entryTime) {
      throw new Error('Entry time is required.');
    }
    this.entryTime = entryTime;
  }

  public close(exitTime: Date): void {
    this.exitTime = ParkingTicket.validateExitTime(this.entryTime, exitTime);
  }

  public calculateTotal(exitTime?: Date): number {
    const finalExit = exitTime ?? this.exitTime;
    const validatedExit = ParkingTicket.validateExitTime(this.entryTime, finalExit);
    const hours = ParkingTicket.hoursRoundedUp(this.entryTime, validatedExit);
    return hours * ParkingTicket.HOURLY_RATE;
  }

  public getEntryTime(): Date {
    return new Date(this.entryTime.getTime());
  }

  public getExitTime(): Date | undefined {
    return this.exitTime ? new Date(this.exitTime.getTime()) : undefined;
  }

  private static validateExitTime(entry: Date, exitTime?: Date): Date {
    if (!exitTime) {
      throw new Error('Exit time is required.');
    }
    if (exitTime.getTime() <= entry.getTime()) {
      throw new Error('Exit time must be after entry time.');
    }
    return exitTime;
  }

  private static hoursRoundedUp(entry: Date, exitTime: Date): number {
    const minutes = (exitTime.getTime() - entry.getTime()) / 60000;
    return Math.ceil(minutes / 60);
  }
}

File: Aula01-introducao-poo/java/ParkingTicket.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.Duration;
import java.time.LocalDateTime;

public class ParkingTicket {
    private static final BigDecimal DEFAULT_HOURLY_RATE = new BigDecimal("6.00");

    private final LocalDateTime entryTime;
    private final BigDecimal hourlyRate;
    private LocalDateTime exitTime;

    public ParkingTicket(LocalDateTime entryTime) {
        this(entryTime, DEFAULT_HOURLY_RATE);
    }

    public ParkingTicket(LocalDateTime entryTime, BigDecimal hourlyRate) {
        if (entryTime == null) {
            throw new IllegalArgumentException("Entry time is required.");
        }
        if (hourlyRate == null || hourlyRate.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Hourly rate must be positive.");
        }
        this.entryTime = entryTime;
        this.hourlyRate = hourlyRate;
    }

    public void close(LocalDateTime exitTime) {
        this.exitTime = validateExitTime(exitTime);
    }

    public BigDecimal calculateTotal() {
        return calculateTotal(this.exitTime);
    }

    public BigDecimal calculateTotal(LocalDateTime exitTime) {
        LocalDateTime validatedExit = validateExitTime(exitTime);
        long minutes = Duration.between(entryTime, validatedExit).toMinutes();
        long hours = (minutes + 59) / 60;
        return hourlyRate.multiply(BigDecimal.valueOf(hours))
                .setScale(2, RoundingMode.HALF_UP);
    }

    public LocalDateTime getEntryTime() {
        return entryTime;
    }

    public LocalDateTime getExitTime() {
        return exitTime;
    }

    private LocalDateTime validateExitTime(LocalDateTime exitTime) {
        if (exitTime == null) {
            throw new IllegalArgumentException("Exit time is required.");
        }
        if (!exitTime.isAfter(entryTime)) {
            throw new IllegalArgumentException("Exit time must be after entry time.");
        }
        return exitTime;
    }
}
